File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    let n = heights.length, ans = 0;
    st = [];

    for(let i = 0; i<=n; i++) {
        while(st.length > 0 && (i == n || heights[st[st.length - 1]] > heights[i])) {
            let h = heights[st.pop()];
            let l = st.length == 0 ? i : i - st[st.length - 1] - 1;
            ans = Math.max(ans, l * h);
        }
        st.push(i);
    }

    return ans;
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        int n = heights.size();
        int ans = 0;
        stack<int> st;

        for(int i = 0; i<n; i++) {
            while(!st.empty() && heights[st.top()] > heights[i]) {
                int h = heights[st.top()];
                st.pop();
                if(st.empty()) {
                    ans = max(ans, h*i);
                }
                else {
                    int l = i - st.top() - 1;
                    ans = max(ans, h * l);
                }
            }
            st.push(i);
        }

        return ans;
    }
};